Docusaurus CLI throws when no subcommand is provided and when not using conventional config name

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

Attemt to specify custom location for docusaurus config fails with same error as if flag not provided at all, in fact, docusaurus seems not to take any flags at all, as `--help` flag (dont know if it exists at all) dont take any effect as well.



### Reproducible demo

_No response_

### Steps to reproduce

1. define relative config path as `--config <path>`
2. get an error that config not found

### Expected behavior

`--config` flag takes effect

### Actual behavior

Config file not read from specified path, though it exists

```powershell
> ls ./.docusaurus/docusaurus.config.ts

Directory: ....\.docusaurus

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          17.12.2023    23:27            811 docusaurus.config.ts

> yarn docusaurus --config ./.docusaurus/docusaurus.config.ts
yarn run v1.22.21
$ ....\node_modules\.bin\docusaurus --config ./.docusaurus/docusaurus.config.ts
[ERROR] No config file found.
[INFO] Expected one of:
- docusaurus.config.ts
- docusaurus.config.mts
- docusaurus.config.cts
- docusaurus.config.js
- docusaurus.config.mjs
- docusaurus.config.cjs
You can provide a custom config path with the `--config` option.
....\node_modules\@docusaurus\core\lib\server\config.js:24
throw new Error();
^

Error
at findConfig (....\node_modules\@docusaurus\core\lib\server\config.js:24:15)
at async loadSiteConfig (....\node_modules\@docusaurus\core\lib\server\config.js:31:11)
at async loadContext (....\node_modules\@docusaurus\core\lib\server\index.js:31:63)
at async externalCommand (....\node_modules\@docusaurus\core\lib\commands\external.js:16:21)
at async file:///..../node_modules/@docusaurus/core/bin/docusaurus.mjs:241:3

Node.js v20.10.0
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Your environment

- Docusaurus version used: 3.0.1
- Environment name and version (e.g. Chrome 89, Node.js 16.4): Nodejs v20.10.0
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): Windows 11


### Self-service

- [ ] I'd be willing to fix this bug myself.

--- a/packages/docusaurus/bin/docusaurus.mjs
+++ b/packages/docusaurus/bin/docusaurus.mjs
@@ -218,6 +218,9 @@ cli.arguments('<command>').action((cmd) => {
   logger.error`    Unknown command name=${cmd}.`;
 });
 
+// === The above is the commander configuration ===
+// They don't start any code execution yet until cli.parse() is called below
+
 /**
  * @param {string | undefined} command
  */
@@ -237,12 +240,29 @@ function isInternalCommand(command) {
   );
 }
 
-if (!isInternalCommand(process.argv.slice(2)[0])) {
-  await externalCommand(cli);
+// process.argv always looks like this:
+// [
+//   '/path/to/node',
+//   '/path/to/docusaurus.mjs',
+//   '<subcommand>',
+//   ...subcommandArgs
+// ]
+
+// There is no subcommand
+// TODO: can we use commander to handle this case?
+if (process.argv.length < 3 || process.argv[2]?.startsWith('--')) {
+  cli.outputHelp();
+  process.exit(1);
 }
 
-if (!process.argv.slice(2).length) {
-  cli.outputHelp();
+// There is an unrecognized subcommand
+// Let plugins extend the CLI before parsing
+if (!isInternalCommand(process.argv[2])) {
+  // TODO: in this step, we must assume default site structure because there's
+  // no way to know the siteDir/config yet. Maybe the root cli should be
+  // responsible for parsing these arguments?
+  // https://github.com/facebook/docusaurus/issues/8903
+  await externalCommand(cli);
 }
 
 cli.parse(process.argv);

